### PR TITLE
Auto detect metrics and their direction

### PR DESCRIPTION
**Pull request by fritschldwg**
# Auto detect metrics and their direction

Closes: #2486

## Description

**Descriptive pull request text**, answering:
  - The `direction` of metrics gets calulated as part of the analysis and is added as a new field to the JSON schema
  - The direction of metrics is used in the visualization to invert area, height and color when map is loaded

## Definition of Done

A PR is only ready for merge once all the following acceptance criteria are fulfilled:
- [x] Changes have been manually tested
- [x] All TODOs related to this PR have been closed
- [x] There are automated tests for newly written code and bug fixes
- [x] All bugs discovered while working on this PR have been submitted as issues (if not already an open issue)
- [x] Documentation (GH-pages, analysis/visualization READMEs, parser READMEs, --help, etc.) has been updated (almost always necessary except for bug fixes)
- [x] CHANGELOG.md has been updated

### CODE REVIEW

File: analysis/parser/RawTextParser/src/main/kotlin/de/maibornwolff/codecharta/parser/rawtextparser/ProjectGenerator.kt [hunk 1]
```diff
@@ -8,9 +8,9 @@ import de.maibornwolff.codecharta.model.ProjectBuilder
 
 class ProjectGenerator(private val projectBuilder: ProjectBuilder = ProjectBuilder()) {
 
-    fun generate(projectMetrics: ProjectMetrics, pipedProject: Project?): Project {
+    fun generate(projectMetrics: ProjectMetrics, maxIndentLevel: Int, pipedProject: Project?): Project {
```
Comment (Nereboss) [inline analysis/parser/RawTextParser/src/main/kotlin/de/maibornwolff/codecharta/parser/rawtextparser/ProjectGenerator.kt:11]: I think the maxIndentLevel should have a default value here so the function can be called without it

File: gh-pages/_docs/04-04-svnlogparser.md [hunk 2]
```diff
@@ -7,16 +7,18 @@ title: "SVN Log Parser"
...
+| ----------------------------- | ------------------------------------------------------------------------------------- |
+| `age_in_weeks`                | age of the file in weeks                                                              |
+| `authors`                     | authors that have worked on a file                                                    |
+| `number_of_authors`           | number of authors with commits                                                        |
+| `number_of_commits`           | total number of commits                                                               |
+| `number_of_renames`           | total number of renames                                                               |
+| `range_of_weeks_with_commits` | average number of weeks between commits                                               |
+| `successive_weeks_of_commits` | number of successive weeks in which the file was included in a commit                 |
+| `weeks_with_commits`          | weeks with commits                                                                    |
```
Comment (Nereboss) [inline gh-pages/_docs/04-04-svnlogparser.md:19]: This description can be improved. For example "Number of weeks in which the file was included in a commit"

File: gh-pages/_docs/04-12-gitlogparser.md [hunk 3]
```diff
@@ -18,16 +18,17 @@ Things to note:
...
+| Metric                          | Description                                                                           |
+| ------------------------------- | ------------------------------------------------------------------------------------- |
+| `age_in_weeks`                  | age of the file in weeks                                                              |
+| `number_of_authors`             | number of authors with commits                                                        |
+| `number_of_commits`             | total number of commits                                                               |
+| `number_of_renames`             | total number of renames                                                               |
+| `range_of_weeks_with_commits`   | week Range of Commits                                                                 |
+| `successive_weeks_with_commits` | successive Weeks with Commits                                                         |
+| `weeks_with_commits`            | weeks with commits                                                                    |
```
Comment (Nereboss) [inline gh-pages/_docs/04-12-gitlogparser.md:29]: Description can be improved (see last comment)

File: visualization/CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/)
     <img width="300" alt="color-metric-diagram" src="https://github.com/MaibornWolff/codecharta/assets/149768491/0b0e4e6f-e0c2-49b1-b9e8-2d69002338f0">
     </br>
 -   Persist uploaded files and applied configuration accross page-reloads [#3486](https://github.com/MaibornWolff/codecharta/pull/3486)
+-   Auto detect direction of metrics [#3518](https://github.com/MaibornWolff/codecharta/pull/3518)
```
Comment (Nereboss) [inline visualization/CHANGELOG.md:17]: Could explain a bit more what this means, e.g. "Auto detect direction of metrics, meaning metrics where lower is better are now displayed differently than metrics where higher is better"

**Review by Nereboss — COMMENTED** [review]
some smaller things in the individual comments but we should think about whether storing the direction as an int with -1 and 1 is good or if we rather refactor it to be invert_direction and be boolean, so its more readable

**Comment by fritschldwg** [comment]
> some smaller things in the individual comments but we should think about whether storing the direction as an int with -1 and 1 is good or if we rather refactor it to be invert_direction and be boolean, so its more readable

After thinking it over, changing the direction property to a boolean seems like a lot of work for not much benefit. Sonar uses the same `direction` property, with -1 for standard metrics and 1 for inverted metrics, just like us. Sticking with this makes things easier for everyone, especially those who already know how Sonar works

File: analysis/import/CSVImporter/README.md [hunk 5]
```diff
@@ -16,6 +16,23 @@ _Conventions for csv input:_
...
+| --------------------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
+| `loc`                             | Lines of code including empty lines and comments                                                                                  |
+| `rloc`                            | Number of physical lines that contain at least one character which is neither a whitespace nor a tabulation nor part of a comment |
+| `classes`                         | Number of classes                                                                                                                 |
+| `functions_per_class`             | Number of functions per class                                                                                                     |
+| `average_statements_per_function` | Average number of statements per method                                                                                           |
+| `max_function_mcc`                | Maximum cyclic complexity based on paths through a function by McCabe                                                             |
```
[the remark this answers is not in the document]
Comment (fritschldwg) [reply analysis/import/CSVImporter/README.md:28]: Throughout the docs there a plenty of links to both sonar's definition as well as wikipedia's definition of the cyclomatic complexity.

File: analysis/import/CSVImporter/src/test/kotlin/de/maibornwolff/codecharta/importer/csv/CSVProjectBuilderTest.kt [hunk 6]
```diff
@@ -14,111 +15,130 @@ class CSVProjectBuilderTest {
...
         val childName = project.rootNode.children.map { it.name }
 
+        // then
         assertEquals(childName[0], name)
     }
 
     @Test
-    fun `adding valid csv with windows line breaks`() {
-        val csvProjectBuilder = CSVProjectBuilder('\\', ',')
+    fun `it should read the csv content correctly when csv contains windows line-breaks`() {
```
[the remark this answers is not in the document]
Comment (fritschldwg) [reply analysis/import/CSVImporter/src/test/kotlin/de/maibornwolff/codecharta/importer/csv/CSVProjectBuilderTest.kt:48]: Both Unix and OS X use linefeed, which is used in every test except the one for windows-linebreaks.

File: analysis/import/MetricGardenerImporter/README.md [hunk 7]
```diff
@@ -31,12 +31,14 @@ If you want to execute the `metric-gardener`, either with your own installation
...
 
--   mcc
--   functions
--   classes
--   lines_of_code
--   comment_lines
--   real_lines_of_code
+| Metric               | Description                                                                                                                       |
+| -------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
+| `mcc`                | Maximum cyclic complexity based on paths through the code by McCabe                                                               |
```
[the remark this answers is not in the document]
Comment (fritschldwg) [reply analysis/import/MetricGardenerImporter/README.md:36]: See a few comments above.

File: analysis/model/src/main/kotlin/de/maibornwolff/codecharta/model/ProjectBuilder.kt [hunk 8]
```diff
@@ -87,11 +85,76 @@ open class ProjectBuilder(
...
 
+    private fun addAttributeDescriptorWithEstimatedDirection(
+            nodeAttributeName: String,
+            complementedAttributeDescriptors: MutableMap<String, AttributeDescriptor>
+                                                            ) {
+        complementedAttributeDescriptors[nodeAttributeName] =
+                AttributeDescriptor(title = nodeAttributeName, direction = estimateDirection(nodeAttributeName))
+    }
+
+    private fun estimateDirection(nodeAttributeName: String): Int {
```
[the remark this answers is not in the document]
Comment (fritschldwg) [reply analysis/model/src/main/kotlin/de/maibornwolff/codecharta/model/ProjectBuilder.kt:117]: We have a list of directions for metrics we know. For every metric we don't know (eg imported through the csv-importer) we do an estimation. On this note - please review the actual functionality of this ticket (wether that estimation does its job appropriately).

File: analysis/parser/RawTextParser/src/main/kotlin/de/maibornwolff/codecharta/parser/rawtextparser/ProjectGenerator.kt (hunk 1 above)
Comment (fritschldwg) [reply analysis/parser/RawTextParser/src/main/kotlin/de/maibornwolff/codecharta/parser/rawtextparser/ProjectGenerator.kt:11]: The ProjectGenerator is only called once within the RawTextParser, with the maximum indentation level set. - there's no need to make this an optional parameter.

**Review by Nereboss — APPROVED** [review]
LGTM!

**Pull request merged**